content/usgs-packages/geoknife_Assessment.Rmd

```diff
@@ -46,4 +47,70 @@ Before starting the exercises, you should make sure that the `geoknife` package
 library(geoknife)
 ```
 
+## Exercise 1
 
+*How many GDP data sets are related to sea level rise? Hint: `grep`.*
+
+```{r addToggle='ON'}
+# First, you need to query for all web data
+all_webdata <- query("webdata")
+all_titles <- title(all_webdata)
+all_abstracts <- abstract(all_webdata)
+
+# Then start sleuthing using `grep`
+keyword_str <- "sea level rise|sea level|sea level|sea rise"
+index_t <- grep(keyword_str, all_titles, ignore.case = TRUE)
+index_a <- grep(keyword_str, all_abstracts, ignore.case = TRUE)
+index_both <- unique(c(index_t, index_a))
+
+# Look at the titles and abstracts of datasets that match your criteria
+length(index_both)
+all_titles[index_both]
+all_abstracts[index_both]
+```
+
+## Exercise 2
+
+*What variables are in the "University of Idaho Daily Meteorological data for continental US" dataset? Also, choose one variable and determine the range of dates. Hint: create a geoknife fabric first.*
+
+```{r addToggle='ON'}
+# Use the all_webdata object from Exercise 1 to create the fabric
+us_meterology <- webdata(all_webdata["University of Idaho Daily Meteorological data for continental US"])
+
+# Now use query to see what variables are available
+metero_vars <- query(us_meterology, "variables")
+metero_vars
+
+# Let's pick the fourth variable to look at a date range
+# To determine the times available, you must add the variable to the fabric
+variables(us_meterology) <- metero_vars[4]
+query(us_meterology, "times")
+```
+
+## Exercise 3
+
+*What was the average maximum air temperature in Texas on July 4, 2007? Use the dataset titled "TopoWx: Topoclimatic Daily Air Temperature Dataset for the Conterminous United States", which has a maximum temperature variable, tmax. Hint: you will need all three pieces - stencil, fabric, and knife.*
+
+```{r addToggle='ON'}
+# Setup the maximum air temp fabric using the URL in all_webdata from the previous lesson
+airtemp_title <- "TopoWx: Topoclimatic Daily Air Temperature Dataset for the Conterminous United States"
+airtemp_url <-  url(all_webdata[airtemp_title])
+airtemp_fabric <- webdata(list(
+  url = airtemp_url,
+  variables = "tmax",
+  times = as.POSIXct(c("2007-07-04", "2007-07-04"), tz = "UTC")
```

Comment:
The reason this is giving the wrong date is because this range ends up setting start_date to `2007-07-04T00:00:00Z` as well as end_date.  Because the dataset centers on 12:00:00 to indicate daily values, the range doesn't overlap.  It does however capture the previous day because the dates are interpreted as lower end inclusive, upper end exclusive.  This oddly makes it hard to not get data for the 3rd.  This is quite possibly an issue with time handling in the GDP as it seems to be assuming days go from 12:00 to 12:00.  So to get it to return just the 4th, the start time needs to be after `2007-07-04T12:00:00Z` and the end time needs to be before `2007-07-05T12:00:00Z` (I chose 23:59:59).